File: mac/Sources/VibeBoardKit/Assets/AssetPixels.swift
import Foundation

public enum AssetConversionError: Error, Equatable, Sendable {
    case invalidDimensions
    case sourceLimitExceeded
    case decodedMemoryLimitExceeded
    case invalidPixelData
    case unsupportedSource
    case decodeFailed
    case invalidFrameDuration
}

public struct AssetRGB888: Equatable, Sendable {
    public let red: UInt8
    public let green: UInt8
    public let blue: UInt8

    public init(red: UInt8, green: UInt8, blue: UInt8) {
        self.red = red
        self.green = green
        self.blue = blue
    }

    public init(packed: UInt32) throws {
        guard packed <= 0xFF_FFFF else { throw AssetConversionError.invalidPixelData }
        red = UInt8((packed >> 16) & 0xff)
        green = UInt8((packed >> 8) & 0xff)
        blue = UInt8(packed & 0xff)
    }

    public var packed: UInt32 { UInt32(red) << 16 | UInt32(green) << 8 | UInt32(blue) }
}

public struct AssetRGBA8: Equatable, Sendable {
    public let red: UInt8
    public let green: UInt8
    public let blue: UInt8
    public let alpha: UInt8

    public init(red: UInt8, green: UInt8, blue: UInt8, alpha: UInt8 = 255) {
        self.red = red
        self.green = green
        self.blue = blue
        self.alpha = alpha
    }
}

public struct AssetRaster: Equatable, Sendable {
    public let width: Int
    public let height: Int
    public let pixels: [AssetRGBA8]

    public init(width: Int, height: Int, pixels: [AssetRGBA8]) throws {
        let (count, overflow) = width.multipliedReportingOverflow(by: height)
        guard width > 0, height > 0, !overflow, pixels.count == count else {
            throw AssetConversionError.invalidDimensions
        }
        self.width = width
        self.height = height
        self.pixels = pixels
    }
}

public struct AssetConversionLimits: Equatable, Sendable {
    public let maxSourceWidth: Int
    public let maxSourceHeight: Int
    public let maxSourcePixels: Int
    public let maxFrames: Int
    public let maxDecodedBytes: Int

    public init(
        maxSourceWidth: Int = 8_192,
        maxSourceHeight: Int = 8_192,
        maxSourcePixels: Int = 16_777_216,
        maxFrames: Int = 256,
        maxDecodedBytes: Int = 128 * 1_024 * 1_024
    ) {
        self.maxSourceWidth = maxSourceWidth
        self.maxSourceHeight = maxSourceHeight
        self.maxSourcePixels = maxSourcePixels
        self.maxFrames = maxFrames
        self.maxDecodedBytes = maxDecodedBytes
    }

    public func validate(width: Int, height: Int, frames: Int = 1, bytesPerPixel: Int = 4) throws {
        guard width > 0, height > 0, width <= maxSourceWidth, height <= maxSourceHeight,
              frames > 0, frames <= maxFrames else { throw AssetConversionError.sourceLimitExceeded }
        let (pixels, pixelOverflow) = width.multipliedReportingOverflow(by: height)
        let (allPixels, frameOverflow) = pixels.multipliedReportingOverflow(by: frames)
        let (bytes, byteOverflow) = allPixels.multipliedReportingOverflow(by: bytesPerPixel)
        guard !pixelOverflow, !frameOverflow, !byteOverflow,
              pixels <= maxSourcePixels, bytes <= maxDecodedBytes else {
            throw AssetConversionError.decodedMemoryLimitExceeded
        }
    }
}

public enum AssetEXIFOrientation: UInt8, Equatable, Sendable {
    case up = 1, upMirrored, down, downMirrored, leftMirrored, right, rightMirrored, left
}

public enum AssetFit: String, Equatable, Sendable {
    case contain, cover, stretch, center
}

public enum AssetPixelConverter {
    public static func applyOrientation(_ source: AssetRaster, _ orientation: AssetEXIFOrientation) throws -> AssetRaster {
        let swapsAxes = orientation == .leftMirrored || orientation == .right || orientation == .rightMirrored || orientation == .left
        let outputWidth = swapsAxes ? source.height : source.width
        let outputHeight = swapsAxes ? source.width : source.height
        var output = Array(repeating: AssetRGBA8(red: 0, green: 0, blue: 0, alpha: 0), count: outputWidth * outputHeight)
        for y in 0..<outputHeight {
            for x in 0..<outputWidth {
                let sourcePoint: (Int, Int)
                switch orientation {
                case .up: sourcePoint = (x, y)
                case .upMirrored: sourcePoint = (source.width - 1 - x, y)
                case .down: sourcePoint = (source.width - 1 - x, source.height - 1 - y)
                case .downMirrored: sourcePoint = (x, source.height - 1 - y)
                case .leftMirrored: sourcePoint = (y, x)
                case .right: sourcePoint = (y, source.height - 1 - x)
                case .rightMirrored: sourcePoint = (source.width - 1 - y, source.height - 1 - x)
                case .left: sourcePoint = (source.width - 1 - y, x)
                }
                output[y * outputWidth + x] = source.pixels[sourcePoint.1 * source.width + sourcePoint.0]
            }
        }
        return try AssetRaster(width: outputWidth, height: outputHeight, pixels: output)
    }

    public static func convert(
        _ source: AssetRaster,
        width targetWidth: Int = 428,
        height targetHeight: Int = 142,
        fit: AssetFit,
        background: AssetRGB888
    ) throws -> [UInt16] {
        guard targetWidth > 0, targetHeight > 0, targetWidth <= 428, targetHeight <= 142 else {
            throw AssetConversionError.invalidDimensions
        }
        let placement = try placement(sourceWidth: source.width, sourceHeight: source.height, targetWidth: targetWidth, targetHeight: targetHeight, fit: fit)
        var output = Array(repeating: rgb565(background), count: targetWidth * targetHeight)
        for y in 0..<placement.height {
            let targetY = placement.y + y
            guard targetY >= 0, targetY < targetHeight else { continue }
            for x in 0..<placement.width {
                let targetX = placement.x + x
                guard targetX >= 0, targetX < targetWidth else { continue }
                let rgba = bilinear(source, x: x, y: y, width: placement.width, height: placement.height)
                output[targetY * targetWidth + targetX] = rgb565(composite(rgba, over: background))
            }
        }
        return output
    }

    public static func rgb565(_ pixel: AssetRGB888) -> UInt16 {
        (UInt16(pixel.red >> 3) << 11) | (UInt16(pixel.green >> 2) << 5) | UInt16(pixel.blue >> 3)
    }

    public static func composite(_ source: AssetRGBA8, over background: AssetRGB888) -> AssetRGB888 {
        let alpha = UInt32(source.alpha)
        func channel(_ foreground: UInt8, _ destination: UInt8) -> UInt8 {
            let value = UInt32(foreground) * alpha + UInt32(destination) * (255 - alpha) + 127
            return UInt8(value / 255)
        }
        return AssetRGB888(
            red: channel(source.red, background.red),
            green: channel(source.green, background.green),
            blue: channel(source.blue, background.blue)
        )
    }

    private struct Placement { let x: Int; let y: Int; let width: Int; let height: Int }

    private static func placement(sourceWidth: Int, sourceHeight: Int, targetWidth: Int, targetHeight: Int, fit: AssetFit) throws -> Placement {
        guard sourceWidth > 0, sourceHeight > 0 else { throw AssetConversionError.invalidDimensions }
        switch fit {
        case .stretch:
            return Placement(x: 0, y: 0, width: targetWidth, height: targetHeight)
        case .center:
            return Placement(x: (targetWidth - sourceWidth) / 2, y: (targetHeight - sourceHeight) / 2, width: sourceWidth, height: sourceHeight)
        case .contain, .cover:
            let widthLimited = Int64(targetWidth) * Int64(sourceHeight)
            let heightLimited = Int64(targetHeight) * Int64(sourceWidth)
            let useWidth = fit == .contain ? widthLimited <= heightLimited : widthLimited >= heightLimited
            let width: Int
            let height: Int
            if useWidth {
                width = targetWidth
                height = max(1, Int((Int64(sourceHeight) * Int64(targetWidth)) / Int64(sourceWidth)))
            } else {
                height = targetHeight
                width = max(1, Int((Int64(sourceWidth) * Int64(targetHeight)) / Int64(sourceHeight)))
            }
            return Placement(x: (targetWidth - width) / 2, y: (targetHeight - height) / 2, width: width, height: height)
        }
    }

    private static func bilinear(_ source: AssetRaster, x: Int, y: Int, width: Int, height: Int) -> AssetRGBA8 {
        let xAxis = axis(position: x, source: source.width, destination: width)
        let yAxis = axis(position: y, source: source.height, destination: height)
        let p00 = source.pixels[yAxis.low * source.width + xAxis.low]
        let p10 = source.pixels[yAxis.low * source.width + xAxis.high]
        let p01 = source.pixels[yAxis.high * source.width + xAxis.low]
        let p11 = source.pixels[yAxis.high * source.width + xAxis.high]
        func sample(_ keyPath: KeyPath<AssetRGBA8, UInt8>) -> UInt8 {
            let top = UInt64(p00[keyPath: keyPath]) * UInt64(xAxis.denominator - xAxis.remainder) + UInt64(p10[keyPath: keyPath]) * UInt64(xAxis.remainder)
            let bottom = UInt64(p01[keyPath: keyPath]) * UInt64(xAxis.denominator - xAxis.remainder) + UInt64(p11[keyPath: keyPath]) * UInt64(xAxis.remainder)
            let denominator = UInt64(xAxis.denominator) * UInt64(yAxis.denominator)
            let numerator = top * UInt64(yAxis.denominator - yAxis.remainder) + bottom * UInt64(yAxis.remainder)
            return UInt8(min(255, (numerator + denominator / 2) / denominator))
        }
        return AssetRGBA8(red: sample(\.red), green: sample(\.green), blue: sample(\.blue), alpha: sample(\.alpha))
    }

    private struct Axis { let low: Int; let high: Int; let remainder: Int; let denominator: Int }

    private static func axis(position: Int, source: Int, destination: Int) -> Axis {
        let denominator = 2 * destination
        let numerator = (2 * position + 1) * source - destination
        if numerator <= 0 { return Axis(low: 0, high: 0, remainder: 0, denominator: denominator) }
        let maximum = (source - 1) * denominator
        if numerator >= maximum { return Axis(low: source - 1, high: source - 1, remainder: 0, denominator: denominator) }
        let low = numerator / denominator
        return Axis(low: low, high: low + 1, remainder: numerator % denominator, denominator: denominator)
    }
}

File: win/src/VibeBoardKit/Assets/AssetPixels.cs
using System;
using System.Collections.Generic;

namespace VibeBoardKit.Assets;

/// <summary>
/// RGB888 pixel.
/// </summary>
public readonly struct AssetRGB888
{
    public byte Red { get; }
    public byte Green { get; }
    public byte Blue { get; }
    public AssetRGB888(byte r, byte g, byte b) { Red = r; Green = g; Blue = b; }
}

/// <summary>
/// RGBA8 pixel.
/// </summary>
public readonly struct AssetRGBA8
{
    public byte Red { get; }
    public byte Green { get; }
    public byte Blue { get; }
    public byte Alpha { get; }
    public AssetRGBA8(byte r, byte g, byte b, byte a)
    { Red = r; Green = g; Blue = b; Alpha = a; }
}

/// <summary>
/// A raster of RGBA8 pixels.
/// </summary>
public sealed class AssetRaster
{
    public int Width { get; }
    public int Height { get; }
    public AssetRGBA8[] Pixels { get; }

    public AssetRaster(int width, int height, AssetRGBA8[] pixels)
    {
        Width = width;
        Height = height;
        Pixels = pixels;
    }
}

/// <summary>
/// Asset conversion limits for source images.
/// </summary>
public sealed class AssetConversionLimits
{
    public int MaxSourceWidth { get; } = 8192;
    public int MaxSourceHeight { get; } = 8192;
    public int MaxSourcePixels { get; } = 16 * 1024 * 1024;
    public int MaxFrames { get; } = 256;
    public long MaxDecodedBytes { get; } = 128 * 1024 * 1024;
}

/// <summary>
/// EXIF orientation values.
/// </summary>
public enum AssetEXIFOrientation : byte
{
    Up = 1,
    UpMirrored = 2,
    Down = 3,
    DownMirrored = 4,
    LeftMirrored = 5,
    Left = 6,
    RightMirrored = 7,
    Right = 8
}

/// <summary>
/// Pixel conversion utilities for asset processing.
/// </summary>
public static class AssetPixelConverter
{
    /// <summary>
    /// Convert an RGBA raster to RGB565 uint16 pixels.
    /// </summary>
    public static ushort[] Convert(AssetRaster raster, int targetWidth, int targetHeight,
        string fit, AssetRGB888 background)
    {
        if (targetWidth < 1 || targetWidth > 428 || targetHeight < 1 || targetHeight > 142)
            throw new ArgumentException("Invalid target dimensions");

        // Composite RGBA over background
        var composited = CompositeOverBackground(raster, background);

        // Scale to target dimensions
        var scaled = fit switch
        {
            "contain" => ScaleContain(composited, raster.Width, raster.Height, targetWidth, targetHeight),
            "cover" => ScaleCover(composited, raster.Width, raster.Height, targetWidth, targetHeight),
            "stretch" => ScaleStretch(composited, raster.Width, raster.Height, targetWidth, targetHeight),
            "center" => ScaleCenter(composited, raster.Width, raster.Height, targetWidth, targetHeight),
            _ => ScaleContain(composited, raster.Width, raster.Height, targetWidth, targetHeight)
        };

        // Convert to RGB565
        var result = new ushort[targetWidth * targetHeight];
        for (int i = 0; i < scaled.Length; i++)
        {
            var p = scaled[i];
            result[i] = RGB565(p.Red, p.Green, p.Blue);
        }
        return result;
    }

    public static ushort RGB565(byte r, byte g, byte b)
    {
        return (ushort)(((r >> 3) << 11) | ((g >> 2) << 5) | (b >> 3));
    }

    /// <summary>
    /// Apply EXIF orientation to a raster.
    /// </summary>
    public static AssetRaster ApplyOrientation(AssetRaster raster, AssetEXIFOrientation orientation)
    {
        if (orientation == AssetEXIFOrientation.Up)
            return raster;

        int w = raster.Width;
        int h = raster.Height;
        var src = raster.Pixels;

        AssetRGBA8[] dst;
        int newW, newH;

        switch (orientation)
        {
            case AssetEXIFOrientation.UpMirrored:
                dst = new AssetRGBA8[src.Length];
                newW = w; newH = h;
                for (int y = 0; y < h; y++)
                    for (int x = 0; x < w; x++)
                        dst[y * w + (w - 1 - x)] = src[y * w + x];
                break;

            case AssetEXIFOrientation.Down:
                dst = new AssetRGBA8[src.Length];
                newW = w; newH = h;
                for (int y = 0; y < h; y++)
                    for (int x = 0; x < w; x++)
                        dst[(h - 1 - y) * w + (w - 1 - x)] = src[y * w + x];
                break;

            case AssetEXIFOrientation.DownMirrored:
                dst = new AssetRGBA8[src.Length];
                newW = w; newH = h;
                for (int y = 0; y < h; y++)
                    for (int x = 0; x < w; x++)
                        dst[(h - 1 - y) * w + x] = src[y * w + x];
                break;

            case AssetEXIFOrientation.Left:
            case AssetEXIFOrientation.RightMirrored:
                dst = new AssetRGBA8[src.Length];
                newW = h; newH = w;
                for (int y = 0; y < h; y++)
                    for (int x = 0; x < w; x++)
                    {
                        int nx = orientation == AssetEXIFOrientation.Left ? (h - 1 - y) : y;
                        int ny = orientation == AssetEXIFOrientation.Left ? x : (w - 1 - x);
                        dst[ny * newW + nx] = src[y * w + x];
                    }
                break;

            case AssetEXIFOrientation.Right:
            case AssetEXIFOrientation.LeftMirrored:
                dst = new AssetRGBA8[src.Length];
                newW = h; newH = w;
                for (int y = 0; y < h; y++)
                    for (int x = 0; x < w; x++)
                    {
                        int nx = orientation == AssetEXIFOrientation.Right ? y : (h - 1 - y);
                        int ny = orientation == AssetEXIFOrientation.Right ? (w - 1 - x) : x;
                        dst[ny * newW + nx] = src[y * w + x];
                    }
                break;

            default:
                return raster;
        }

        return new AssetRaster(newW, newH, dst);
    }

    private static AssetRGB888[] CompositeOverBackground(AssetRaster raster, AssetRGB888 background)
    {
        var result = new AssetRGB888[raster.Pixels.Length];
        for (int i = 0; i < raster.Pixels.Length; i++)
        {
            var p = raster.Pixels[i];
            if (p.Alpha == 255)
            {
                result[i] = new AssetRGB888(p.Red, p.Green, p.Blue);
            }
            else if (p.Alpha == 0)
            {
                result[i] = background;
            }
            else
            {
                float a = p.Alpha / 255f;
                byte r = (byte)(p.Red * a + background.Red * (1 - a));
                byte g = (byte)(p.Green * a + background.Green * (1 - a));
                byte b = (byte)(p.Blue * a + background.Blue * (1 - a));
                result[i] = new AssetRGB888(r, g, b);
            }
        }
        return result;
    }

    private static AssetRGB888[] ScaleContain(AssetRGB888[] src, int srcW, int srcH, int dstW, int dstH)
    {
        var result = new AssetRGB888[dstW * dstH];
        // Fill with black (background already composited)
        for (int i = 0; i < result.Length; i++)
            result[i] = new AssetRGB888(0, 0, 0);

        float scale = Math.Min((float)dstW / srcW, (float)dstH / srcH);
        int scaledW = (int)(srcW * scale);
        int scaledH = (int)(srcH * scale);
        int offsetX = (dstW - scaledW) / 2;
        int offsetY = (dstH - scaledH) / 2;

        BilinearScale(src, srcW, srcH, result, dstW, dstH, scaledW, scaledH, offsetX, offsetY);
        return result;
    }

    private static AssetRGB888[] ScaleCover(AssetRGB888[] src, int srcW, int srcH, int dstW, int dstH)
    {
        var result = new AssetRGB888[dstW * dstH];
        float scale = Math.Max((float)dstW / srcW, (float)dstH / srcH);
        int scaledW = (int)(srcW * scale);
        int scaledH = (int)(srcH * scale);
        int offsetX = (dstW - scaledW) / 2;
        int offsetY = (dstH - scaledH) / 2;

        BilinearScale(src, srcW, srcH, result, dstW, dstH, scaledW, scaledH, offsetX, offsetY);
        return result;
    }

    private static AssetRGB888[] ScaleStretch(AssetRGB888[] src, int srcW, int srcH, int dstW, int dstH)
    {
        var result = new AssetRGB888[dstW * dstH];
        BilinearScale(src, srcW, srcH, result, dstW, dstH, dstW, dstH, 0, 0);
        return result;
    }

    private static AssetRGB888[] ScaleCenter(AssetRGB888[] src, int srcW, int srcH, int dstW, int dstH)
    {
        var result = new AssetRGB888[dstW * dstH];
        for (int i = 0; i < result.Length; i++)
            result[i] = new AssetRGB888(0, 0, 0);

        int offsetX = (dstW - srcW) / 2;
        int offsetY = (dstH - srcH) / 2;

        for (int y = 0; y < srcH; y++)
        {
            int dy = y + offsetY;
            if (dy < 0 || dy >= dstH) continue;
            for (int x = 0; x < srcW; x++)
            {
                int dx = x + offsetX;
                if (dx < 0 || dx >= dstW) continue;
                result[dy * dstW + dx] = src[y * srcW + x];
            }
        }
        return result;
    }

    private static void BilinearScale(AssetRGB888[] src, int srcW, int srcH,
        AssetRGB888[] dst, int dstW, int dstH,
        int scaledW, int scaledH, int offsetX, int offsetY)
    {
        if (scaledW <= 0 || scaledH <= 0) return;

        float xRatio = (float)srcW / scaledW;
        float yRatio = (float)srcH / scaledH;

        for (int y = 0; y < scaledH; y++)
        {
            int dy = y + offsetY;
            if (dy < 0 || dy >= dstH) continue;

            float sy = y * yRatio;
            int sy0 = (int)sy;
            int sy1 = Math.Min(sy0 + 1, srcH - 1);
            float fy = sy - sy0;

            for (int x = 0; x < scaledW; x++)
            {
                int dx = x + offsetX;
                if (dx < 0 || dx >= dstW) continue;

                float sx = x * xRatio;
                int sx0 = (int)sx;
                int sx1 = Math.Min(sx0 + 1, srcW - 1);
                float fx = sx - sx0;

                var p00 = src[sy0 * srcW + sx0];
                var p01 = src[sy0 * srcW + sx1];
                var p10 = src[sy1 * srcW + sx0];
                var p11 = src[sy1 * srcW + sx1];

                byte r = Blerp(p00.Red, p01.Red, p10.Red, p11.Red, fx, fy);
                byte g = Blerp(p00.Green, p01.Green, p10.Green, p11.Green, fx, fy);
                byte b = Blerp(p00.Blue, p01.Blue, p10.Blue, p11.Blue, fx, fy);

                dst[dy * dstW + dx] = new AssetRGB888(r, g, b);
            }
        }
    }

    private static byte Blerp(byte v00, byte v01, byte v10, byte v11, float fx, float fy)
    {
        float top = v00 * (1 - fx) + v01 * fx;
        float bottom = v10 * (1 - fx) + v11 * fx;
        return (byte)(top * (1 - fy) + bottom * fy);
    }
}
